### PR TITLE
Fix building with crossbuilder

### LINKED ISSUE #46: instructions for setting up, building and testing the project
### problem
I'd like to add an on-screen keyboard layout for an additional language but until now I am not able to compile nor to test the current state of the repo.

So far I tried to
- find the missing packages that qtcreator lists to me but still there are some missing
- use crossbuilder but it ran into problems itself

Still I don't know the intended way of doing this and what I try is also not working. So this situation is quite ... frustating.

### proposal
extend the readme 
- either by instructions on setting up the project for development, building as well as testing it
- or by references that explain those points

### PR DESCRIPTION
**Pull request by jonnius**
Fixes #46.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+Ubuntu Keyboard
+===============
+
+Ubuntu Keyboard is the official QML and C++ based keyboard plugin for Ubuntu Touch. 
+We follow an open source model where the code is available to anyone to branch and hack on.
+
+Updating presage language models
+================================
+
+If the source material used for presage's prediction models is update the databases can be regenerated by running "make lang_db_$lc" within the plugin's "src" directory (where $lc is the language code for the updated language, e.g. "make lang_db_en" for English)
```
Comment (mardy) [inline README.md]: I strongly suggest wrapping lines at 80 chars :-)
Also, to make clear that "presage" is a software name, I'd format it differently; maybe use backticks?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+```
+crossbuilder
+```
+
+See [crossbuilder on github](https://github.com/ubports/crossbuilder) for details.
+
+Useful Links
+============
+Here are some useful links with regards to the Music App development.
```
Comment (mardy) [inline README.md]: Why Music App?

File: debian/rules [hunk 3]
```diff
@@ -17,27 +21,32 @@ override_dh_auto_configure:
...
+        HOME=$(CURDIR)/debian/test-home xvfb-run -a make check
+    endif
 endif
 
 override_dh_clean:
 	rm -rf $(CURDIR)/debian/test-home
 	dh_clean
 
 override_dh_install:
+ifeq (,$(findstring nocheck,$(DEB_BUILD_OPTIONS)))
```
Comment (mardy) [inline debian/rules]: This is not 100% correct: the code here is installing files, not running tests. I think we should be running this in any case. If you do that change I suggest in `debian/control`, you'll see that this works fine.

File: debian/control [hunk 4]
```diff
@@ -17,16 +17,13 @@ Build-Depends: debhelper (>= 9),
                libxml2-utils,
                maliit-framework-dev (>= 0.99.0+git20130615+97e8335-0ubuntu3),
                pkg-config,
-               python3:any,
-               python3-flake8 (>= 2.2.2-1ubuntu4) | python3-flake8:native,
```
Comment (mardy) [inline debian/control:21]: Keep these three lines, but change this one to be just `python3-flake8`. You'll be surprised :-)

**Review by mardy — CHANGES_REQUESTED** [review]
Thanks @jonnius! Looks good, just please see the comments I made, as I think this can be improved.

File: README.md (hunk 2 above)
Comment (jonnius) [reply README.md]: copy paste mistake...

File: README.md (hunk 1 above)
Comment (jonnius) [reply README.md]: This one already was in the README. I tried to improve formatting now.

File: debian/rules (hunk 3 above)
Comment (jonnius) [reply debian/rules]: It looks like it just installs tests...

**Comment by jonnius** [comment]
With the python dependencies I get the dependency issue again:
```
 builddeps:ubuntu-keyboard:armhf : Depends: python3-flake8:armhf but it is not installable
                                   Depends: python3-setuptools:armhf but it is not installable
```
As the `override_dh_install` part only seems to install some tests, I guess we could skip that?

**Comment by mardy** [comment]
We will skip them, if there's no other way around, but yesterday it was working, in my machine. I'll retry with a fresh container.

**Comment by mardy** [comment]
@jonnius, I confirm that I got the same error as you. It looks like in order to install the right dependencies, one has to add `:native` to the python module packages, while `dpkg-checkbuilddeps` wants the packages without it. I wonder if @tsimonq2 has any idea? To me it looks like a bug in either the debian tools, or in the python package metadata.

Anyway, with this diff is I can build the project:
```
diff --git a/debian/control b/debian/control
index 24f441a..73f974f 100644
--- a/debian/control
+++ b/debian/control
@@ -19,8 +19,8 @@ Build-Depends: debhelper (>= 9),
                maliit-framework-dev (>= 0.99.0+git20130615+97e8335-0ubuntu3),
                pkg-config,
                python3:any,
-               python3-flake8 (>= 2.2.2-1ubuntu4) | python3-flake8:native,
-               python3-setuptools,
+               python3-flake8:native | python3-flake8,
+               python3-setuptools:native | python3-setuptools,
                qt5-default,
                qtbase5-dev,
                qtbase5-private-dev,
@@ -35,7 +35,7 @@ Build-Depends: debhelper (>= 9),
                qtdeclarative5-private-dev,
                qml-module-qtmultimedia,
                qml-module-ubuntu-components,
-               xvfb,
+               xvfb <!nocheck>,
                presage,
 Standards-Version: 3.9.5
 Homepage: https://launchpad.net/ubuntu-keyboard
``` 
It is a hack, but we don't lose any functionality with it.

**Comment by mardy** [comment]
OK, after more investigation, it seems like the correct way to install python modules is without the `:native` attribute; howeer that fails because `python3-setuptools` and `python3-flake8` are not marked as `Multi-Arch: foreign`. The first package [has been fixed](https://bugs.launchpad.net/ubuntu/+source/doxyqml/+bug/1585517) in later versions of Ubuntu, and I [filed a bug](https://bugs.launchpad.net/ubuntu/+source/python-flake8/+bug/1795115) for the other one now. But for the time being, the workaround I suggested above is the way to go.

I'll see if I can find a way to automatically detect these situations and run `crossbuilder inst-foreign` on these packages...

**Comment by mardy** [comment]
Last update (hopefully)! :-)

I made a change to crossbuilder where it workarounds the missing `Multi-Arch: foreign` line in some packages. It looks like it works fine, at least with this component :-)
So please try once more with these dependencies
```
    python3-flake8,
    python3-setuptools,
```
and it *should* work (at least, it worked here, with a clean container). :-)

**Comment by jonnius** [comment]
That definitely solved the dependency issue. But I still get an issue with the tests:
```
running install_scripts
# Don't install the other tests
rm debian/tmp/usr/bin/ut_* -f
dh_install -X'*.pyc' --fail-missing
dh_install: ubuntu-keyboard-data missing files: usr/share/maliit/plugins/com/ubuntu/api_tests.qml
dh_install: ubuntu-keyboard-tests missing files: /usr/share/maliit/tests/ubuntu-keyboard
dh_install: missing files, aborting
debian/rules:44: recipe for target 'override_dh_install' failed
make[1]: *** [override_dh_install] Error 2
make[1]: Leaving directory '/home/jonatan/ubuntu-keyboard'
debian/rules:15: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: error: fakeroot debian/rules binary gave error exit status 2
```
I already made the `rm` ignore the missing files. But about that one I have got no idea...

**Comment by mardy** [comment]
It's because of the `notest` in debian/rules. If you remove that check, it will work.

**Comment by jonnius** [comment]
Oh, wow. So tests work again now? Could you give me hint on how to run them? Though I remember the docs saying that with crossbuilder tests are not exported to work...

Anyway. This is ready to merge now. Thanks, again!

**Comment by mardy** [comment]
No, tests don't run with crossbuilder. But they can still be built. With `DEB_BUILD_OPTIONS=nocheck` we tell `dpkg-buildpackage` not to run the tests, because it's likely that they won't pass.
However, they are still built; if you want to avoid building them, you have to use the `notest` option with qmake: that will make the build faster, but it also means that you have to find a way to solve the problem about the missing files. So it might be easier to forget about this last part for the time being.

**Comment by jonnius** [comment]
Thanks for the explanation! Are there any requested changes missing from your review that prevents merging?

**Review by mardy — APPROVED** [review]
Excellent, thanks!

**Pull request merged**